src/main/java/software/amazon/awssdk/crt/s3/S3FinishedResponseContext.java

```diff
@@ -11,48 +11,62 @@ public class S3FinishedResponseContext {
     private final ChecksumAlgorithm checksumAlgorithm;
     private final boolean didValidateChecksum;
 
+    private CrtS3RuntimeException exception;
+
     /*
      * errorCode The CRT error code
      * responseStatus statusCode of the HTTP response
      * errorPayload body of the error response. Can be null if the request completed successfully
      * checksumAlgorithm, the algorithm used to validate the Body, None if not validated
      * didValidateChecksum which is true if the response was validated.
+     * exception CrtS3RuntimeException if any error occurred. May contain the cause of the exception such as a Java exception in a callback.
      */
-    S3FinishedResponseContext(final int errorCode, final int responseStatus, final byte[] errorPayload, final ChecksumAlgorithm checksumAlgorithm, final boolean didValidateChecksum) {
+    S3FinishedResponseContext(final int errorCode, final int responseStatus, final byte[] errorPayload, final ChecksumAlgorithm checksumAlgorithm, final boolean didValidateChecksum, Throwable cause) {
         this.errorCode = errorCode;
         this.responseStatus = responseStatus;
         this.errorPayload = errorPayload;
         this.checksumAlgorithm = checksumAlgorithm;
         this.didValidateChecksum = didValidateChecksum;
+        if (cause != null) {
+            this.exception = new CrtS3RuntimeException(errorCode, responseStatus, errorPayload, cause);
+        } else if (errorCode != 0) {
+            this.exception = new CrtS3RuntimeException(errorCode, responseStatus, errorPayload);
+        }
     }
 
-    public int getErrorCode () {
+    public int getErrorCode() {
         return this.errorCode;
     }
 
     /*
      * If the request didn't receive a response due to a connection
      * failure or some other issue the response status will be 0.
      */
-    public int getResponseStatus () {
+    public int getResponseStatus() {
         return this.responseStatus;
     }
 
     /*
      * In the case of a failed http response get the payload of the response.
      */
-    public byte[] getErrorPayload () {
+    public byte[] getErrorPayload() {
         return this.errorPayload;
     }
 
     /*
      * if no checksum is found, or the request finished with an error the Algorithm will be None,
      * otherwise the algorithm will correspond to the one attached to the object when uploaded.
      */
-    public ChecksumAlgorithm getChecksumAlgorithm () {
+    public ChecksumAlgorithm getChecksumAlgorithm() {
         return this.checksumAlgorithm;
     }
-    public boolean isChecksumValidated () {
+
+    public boolean isChecksumValidated() {
         return this.didValidateChecksum;
     }
+
+    public CrtS3RuntimeException getException() {
```

Comment:
Should we just make this return the (nullable) cause instead of CrtS3RuntimeException? In our current implementation, the CRT client in the SDK just checks the error code and returns our own SdkClientException and we can't change it now. https://github.com/aws/aws-sdk-java-v2/blob/master/services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/S3CrtResponseHandlerAdapter.java#L134

Reply:
The Java SDK can do `exception.errorMessage` instead of parsing the errorCode.